File: src/test/java/javabot/operations/JavadocOperationTest.java
package javabot.operations;

import java.util.Arrays;
import java.util.List;

import javabot.BotEvent;
import javabot.Message;
import javabot.dao.ApiDao;
import javabot.dao.ClazzDao;
import org.springframework.beans.factory.annotation.Autowired;
import org.testng.Assert;
import org.testng.annotations.Test;

/**
 * Created Aug 9, 2007
 *
 * @author <a href="mailto:javabot@cheeseronline.org">cheeser</a>
 */
@Test(dependsOnGroups = {"javadoc"})
public class JavadocOperationTest extends BaseOperationTest {
    @Autowired
    private ApiDao apiDao;
    @Autowired
    private ClazzDao clazzDao;

    public void string() {
        final String response = "http://is.gd/6UoM (JDK)";
        final String errorMessage = "Should have found class";
        final BotOperation operation = getOperation();

        BotEvent event = new BotEvent(CHANNEL, SENDER, LOGIN, HOSTNAME, "javadoc String");
        List<Message> results = operation.handleMessage(event);
        Assert.assertTrue(!results.isEmpty());
        Assert.assertEquals(results.get(0).getMessage(), response, errorMessage);

        event = new BotEvent(CHANNEL, SENDER, LOGIN, HOSTNAME, "javadoc java.lang.String");
        results = operation.handleMessage(event);
        Assert.assertTrue(!results.isEmpty());
        Assert.assertEquals(results.get(0).getMessage(), response, errorMessage);
    }

    public void methods() {
        testOperation("javadoc String.split(String)", "http://is.gd/eOPq (JDK)");
        testOperation("javadoc String.split(java.lang.String)", "http://is.gd/eOPq (JDK)");

        final BotEvent event = new BotEvent(CHANNEL, SENDER, LOGIN, HOSTNAME, "javadoc String.split(*)");
        final Message message = getOperation().handleMessage(event).get(0);
        final String string = message.getMessage();
        final String[] results = string.split(",");
        Assert.assertEquals(results.length, 2);
        final List<String> responses = Arrays.asList("http://is.gd/eOPq (JDK)", "http://is.gd/eOPr (JDK)");
        Assert.assertTrue(responses.contains(results[0].trim()));
        Assert.assertTrue(responses.contains(results[1].trim()));

    }

    @Override
    protected BotOperation createOperation() {
        return new JavadocOperation(getJavabot());
    }
}

File: src/test/kotlin/javabot/operations/JavadocOperationTest.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import javabot.BaseTest
import javabot.dao.JavadocClassDao
import javax.inject.Inject
import org.testng.annotations.BeforeTest
import org.testng.annotations.Test

@Test
class JavadocOperationTest : BaseTest() {
    companion object {
        val STRING_URL = "java.base/java/lang/String.html#"
    }

    @Inject private lateinit var operation: JavadocOperation
    @Inject protected lateinit var javadocClassDao: JavadocClassDao

    @BeforeTest
    fun jdk() {
        loadApi("JDK", version = "11")
    }

    fun constructors() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String(char[])")),
            "$STRING_URL<init>(char[])"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc java.lang.String(char[])")),
            "$STRING_URL<init>(char[])"
        )
    }

    fun methods() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String.split(String)")),
            "${STRING_URL}split(java.lang.String)"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc -jdk String.split(String)")),
            "${STRING_URL}split(java.lang.String)"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.split(java.lang.String)")),
            "${STRING_URL}split(java.lang.String)"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.join(*)")),
            "${STRING_URL}join("
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.split(*)")),
            "${STRING_URL}split(java.lang.String)"
        )
    }

    fun nestedClasses() {
        scanForResponse(
            operation.handleMessage(message("~javadoc Map.Entry")),
            "java/util/Map.Entry.html"
        )
    }

    fun format() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String.format(*)")),
            "java/lang/String.html#format(java.util.Locale,java.lang.String,java.lang.Object...)"
        )
    }

    fun doFinal() {
        scanForResponse(
            operation.handleMessage(message("~javadoc String.valueOf(*)")),
            Sofia.tooManyResults(TEST_USER.nick)
        )
    }

    fun fields() {
        scanForResponse(
            operation.handleMessage(message("~javadoc System.in")),
            "java/lang/System.html#in"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc Integer.MAX_VALUE")),
            "java/lang/Integer.html#MAX_VALUE"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc -jdk System.in")),
            "java/lang/System.html#in"
        )
    }

    fun inherited() {
        scanForResponse(
            operation.handleMessage(message("~javadoc ArrayList.listIterator(*)")),
            "java/util/ArrayList.html#listIterator(int)"
        )
    }

    fun packagePrivate() {
        scanForResponse(
            operation.handleMessage(message("~javadoc ASCII)")),
            "I have no documentation for ASCII"
        )
    }
}
